### ISSUE TITLE
draw_grid issue when the default grid_spacing is larger than a sunpy.map.Map extent 

### ISSUE THREAD

**Issue by hayesla** [sunpy/sunpy#6691]
### Describe the bug

I've come across this issue for a cutout image for which the field of view is smaller than the 15 degrees default grid_spacing, such that it errors when you try to do `my_map.draw_grid()` giving an error thats not informative to the issue. 

### To Reproduce

```python
>>> aiamap = sunpy.map.Map(sunpy.data.sample.AIA_193_CUTOUT01_IMAGE)
>>> bl = SkyCoord(650*u.arcsec, -450*u.arcsec, frame=aiamap.coordinate_frame)
>>> tr = SkyCoord(800*u.arcsec, -300*u.arcsec, frame=aiamap.coordinate_frame)
>>> submap = aiamap.submap(bl, top_right=tr)

>>> submap.plot()
>>> submap.draw_grid()
Error in callback <function install_repl_displayhook.<locals>.post_execute at 0x7fe8850ea4d0> (for post_execute):
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/pyplot.py:137, in install_repl_displayhook.<locals>.post_execute()
    135 def post_execute():
    136     if matplotlib.is_interactive():
--> 137         draw_all()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/_pylab_helpers.py:141, in Gcf.draw_all(cls, force)
    139 for manager in cls.get_all_fig_managers():
    140     if force or manager.canvas.figure.stale:
--> 141         manager.canvas.draw_idle()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/backend_bases.py:2060, in FigureCanvasBase.draw_idle(self, *args, **kwargs)
   2058 if not self._is_idle_drawing:
   2059     with self._idle_draw_cntx():
-> 2060         self.draw(*args, **kwargs)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/backends/backend_agg.py:436, in FigureCanvasAgg.draw(self)
    432 # Acquire a lock on the shared font cache.
    433 with RendererAgg.lock, \
    434      (self.toolbar._wait_cursor_for_draw_cm() if self.toolbar
    435       else nullcontext()):
--> 436     self.figure.draw(self.renderer)
    437     # A GUI class may be need to update a window using this draw, so
    438     # don't forget to call the superclass.
    439     super().draw()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:74, in _finalize_rasterization.<locals>.draw_wrapper(artist, renderer, *args, **kwargs)
     72 @wraps(draw)
     73 def draw_wrapper(artist, renderer, *args, **kwargs):
---> 74     result = draw(artist, renderer, *args, **kwargs)
     75     if renderer._rasterizing:
     76         renderer.stop_rasterizing()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:51, in allow_rasterization.<locals>.draw_wrapper(artist, renderer)
     48     if artist.get_agg_filter() is not None:
     49         renderer.start_filter()
---> 51     return draw(artist, renderer)
     52 finally:
     53     if artist.get_agg_filter() is not None:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/figure.py:2845, in Figure.draw(self, renderer)
   2842         # ValueError can occur when resizing a window.
   2844 self.patch.draw(renderer)
-> 2845 mimage._draw_list_compositing_images(
   2846     renderer, self, artists, self.suppressComposite)
   2848 for sfig in self.subfigs:
   2849     sfig.draw(renderer)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    130 if not_composite or not has_images:
    131     for a in artists:
--> 132         a.draw(renderer)
    133 else:
    134     # Composite any adjacent images together
    135     image_group = []

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:473, in WCSAxes.draw(self, renderer, **kwargs)
    470 # We need to make sure that that frame path is up to date
    471 self.coords.frame._update_patch_path()
--> 473 super().draw(renderer, **kwargs)
    475 self._drawn = True

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:51, in allow_rasterization.<locals>.draw_wrapper(artist, renderer)
     48     if artist.get_agg_filter() is not None:
     49         renderer.start_filter()
---> 51     return draw(artist, renderer)
     52 finally:
     53     if artist.get_agg_filter() is not None:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/axes/_base.py:3091, in _AxesBase.draw(self, renderer)
   3088         a.draw(renderer)
   3089     renderer.stop_rasterizing()
-> 3091 mimage._draw_list_compositing_images(
   3092     renderer, self, artists, self.figure.suppressComposite)
   3094 renderer.close_group('axes')
   3095 self.stale = False

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    130 if not_composite or not has_images:
    131     for a in artists:
--> 132         a.draw(renderer)
    133 else:
    134     # Composite any adjacent images together
    135     image_group = []

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:47, in _WCSAxesArtist.draw(self, renderer, *args, **kwargs)
     46 def draw(self, renderer, *args, **kwargs):
---> 47     self.axes.draw_wcsaxes(renderer)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:423, in WCSAxes.draw_wcsaxes(self, renderer)
    421     coords.frame.update()
    422     for coord in coords:
--> 423         coord._draw_grid(renderer)
    425 for coords in self._all_coords:
    426     # Draw tick labels
    427     for coord in coords:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/coordinate_helpers.py:574, in CoordinateHelper._draw_grid(self, renderer)
    572 else:
    573     if self._grid_type == 'lines':
--> 574         self._update_grid_lines()
    575     else:
    576         self._update_grid_contour()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/coordinate_helpers.py:915, in CoordinateHelper._update_grid_lines(self)
    912 pixel = self.transform.inverted().transform(xy_world)
    914 # Create round-tripped values for checking
--> 915 xy_world_round = self.transform.transform(pixel)
    917 for iw in range(n_coord):
    918     subset = slice(iw * n_samples, (iw + 1) * n_samples)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/transforms.py:1503, in Transform.transform(self, values)
   1500 values = values.reshape((-1, self.input_dims))
   1502 # Transform the values
-> 1503 res = self.transform_affine(self.transform_non_affine(values))
   1505 # Convert the result back to the shape of the input values.
   1506 if ndim == 0:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/transforms.py:2428, in CompositeGenericTransform.transform_non_affine(self, points)
   2426     return self._a.transform_non_affine(points)
   2427 else:
-> 2428     return self._b.transform_non_affine(
   2429                         self._a.transform(points))

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/transforms.py:112, in CoordinateTransform.transform(self, input_coords)
    109     return input_coords
    111 input_coords = input_coords*u.deg
--> 112 x_in, y_in = input_coords[:, 0], input_coords[:, 1]
    114 c_in = SkyCoord(UnitSphericalRepresentation(x_in, y_in),
    115                 frame=self.input_system)
    117 # We often need to transform arrays that contain NaN values, and filtering
    118 # out the NaN values would have a performance hit, so instead we just pass
    119 # on all values and just ignore Numpy warnings

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/units/quantity.py:1206, in Quantity.__getitem__(self, key)
   1203     return self._new_view(self.view(np.ndarray)[key], self.unit[key])
   1205 try:
-> 1206     out = super().__getitem__(key)
   1207 except IndexError:
   1208     # We want zero-dimensional Quantity objects to behave like scalars,
   1209     # so they should raise a TypeError rather than an IndexError.
   1210     if self.isscalar:

IndexError: index 0 is out of bounds for axis 1 with size 0

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/IPython/core/formatters.py:339, in BaseFormatter.__call__(self, obj)
    337     pass
    338 else:
--> 339     return printer(obj)
    340 # Finally look for special method names
    341 method = get_real_method(obj, self.print_method)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/IPython/core/pylabtools.py:151, in print_figure(fig, fmt, bbox_inches, base64, **kwargs)
    148     from matplotlib.backend_bases import FigureCanvasBase
    149     FigureCanvasBase(fig)
--> 151 fig.canvas.print_figure(bytes_io, **kw)
    152 data = bytes_io.getvalue()
    153 if fmt == 'svg':

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/backend_bases.py:2295, in FigureCanvasBase.print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)
   2289     renderer = _get_renderer(
   2290         self.figure,
   2291         functools.partial(
   2292             print_method, orientation=orientation)
   2293     )
   2294     with getattr(renderer, "_draw_disabled", nullcontext)():
-> 2295         self.figure.draw(renderer)
   2297 if bbox_inches:
   2298     if bbox_inches == "tight":

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:74, in _finalize_rasterization.<locals>.draw_wrapper(artist, renderer, *args, **kwargs)
     72 @wraps(draw)
     73 def draw_wrapper(artist, renderer, *args, **kwargs):
---> 74     result = draw(artist, renderer, *args, **kwargs)
     75     if renderer._rasterizing:
     76         renderer.stop_rasterizing()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:51, in allow_rasterization.<locals>.draw_wrapper(artist, renderer)
     48     if artist.get_agg_filter() is not None:
     49         renderer.start_filter()
---> 51     return draw(artist, renderer)
     52 finally:
     53     if artist.get_agg_filter() is not None:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/figure.py:2845, in Figure.draw(self, renderer)
   2842         # ValueError can occur when resizing a window.
   2844 self.patch.draw(renderer)
-> 2845 mimage._draw_list_compositing_images(
   2846     renderer, self, artists, self.suppressComposite)
   2848 for sfig in self.subfigs:
   2849     sfig.draw(renderer)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    130 if not_composite or not has_images:
    131     for a in artists:
--> 132         a.draw(renderer)
    133 else:
    134     # Composite any adjacent images together
    135     image_group = []

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:473, in WCSAxes.draw(self, renderer, **kwargs)
    470 # We need to make sure that that frame path is up to date
    471 self.coords.frame._update_patch_path()
--> 473 super().draw(renderer, **kwargs)
    475 self._drawn = True

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/artist.py:51, in allow_rasterization.<locals>.draw_wrapper(artist, renderer)
     48     if artist.get_agg_filter() is not None:
     49         renderer.start_filter()
---> 51     return draw(artist, renderer)
     52 finally:
     53     if artist.get_agg_filter() is not None:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/axes/_base.py:3091, in _AxesBase.draw(self, renderer)
   3088         a.draw(renderer)
   3089     renderer.stop_rasterizing()
-> 3091 mimage._draw_list_compositing_images(
   3092     renderer, self, artists, self.figure.suppressComposite)
   3094 renderer.close_group('axes')
   3095 self.stale = False

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    130 if not_composite or not has_images:
    131     for a in artists:
--> 132         a.draw(renderer)
    133 else:
    134     # Composite any adjacent images together
    135     image_group = []

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:47, in _WCSAxesArtist.draw(self, renderer, *args, **kwargs)
     46 def draw(self, renderer, *args, **kwargs):
---> 47     self.axes.draw_wcsaxes(renderer)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/core.py:423, in WCSAxes.draw_wcsaxes(self, renderer)
    421     coords.frame.update()
    422     for coord in coords:
--> 423         coord._draw_grid(renderer)
    425 for coords in self._all_coords:
    426     # Draw tick labels
    427     for coord in coords:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/coordinate_helpers.py:574, in CoordinateHelper._draw_grid(self, renderer)
    572 else:
    573     if self._grid_type == 'lines':
--> 574         self._update_grid_lines()
    575     else:
    576         self._update_grid_contour()

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/coordinate_helpers.py:915, in CoordinateHelper._update_grid_lines(self)
    912 pixel = self.transform.inverted().transform(xy_world)
    914 # Create round-tripped values for checking
--> 915 xy_world_round = self.transform.transform(pixel)
    917 for iw in range(n_coord):
    918     subset = slice(iw * n_samples, (iw + 1) * n_samples)

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/transforms.py:1503, in Transform.transform(self, values)
   1500 values = values.reshape((-1, self.input_dims))
   1502 # Transform the values
-> 1503 res = self.transform_affine(self.transform_non_affine(values))
   1505 # Convert the result back to the shape of the input values.
   1506 if ndim == 0:

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/matplotlib/transforms.py:2428, in CompositeGenericTransform.transform_non_affine(self, points)
   2426     return self._a.transform_non_affine(points)
   2427 else:
-> 2428     return self._b.transform_non_affine(
   2429                         self._a.transform(points))

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/visualization/wcsaxes/transforms.py:112, in CoordinateTransform.transform(self, input_coords)
    109     return input_coords
    111 input_coords = input_coords*u.deg
--> 112 x_in, y_in = input_coords[:, 0], input_coords[:, 1]
    114 c_in = SkyCoord(UnitSphericalRepresentation(x_in, y_in),
    115                 frame=self.input_system)
    117 # We often need to transform arrays that contain NaN values, and filtering
    118 # out the NaN values would have a performance hit, so instead we just pass
    119 # on all values and just ignore Numpy warnings

File ~/opt/miniconda3/envs/solar_ana/lib/python3.10/site-packages/astropy/units/quantity.py:1206, in Quantity.__getitem__(self, key)
   1203     return self._new_view(self.view(np.ndarray)[key], self.unit[key])
   1205 try:
-> 1206     out = super().__getitem__(key)
   1207 except IndexError:
   1208     # We want zero-dimensional Quantity objects to behave like scalars,
   1209     # so they should raise a TypeError rather than an IndexError.
   1210     if self.isscalar:

IndexError: index 0 is out of bounds for axis 1 with size 0

<Figure size 640x480 with 1 Axes>

```

this is due to the fact (i suspect) thatthe default `grid_spacing` is `15*u.deg` which when calculated is greater than the field of view its trying to plot on?

if you pass a smaller `grid_spacing` there's no issue:

```python
>>> submap.plot()
>>> submap.draw_grid(grid_spacing=6*u.deg)
![test](https://user-images.githubusercontent.com/4620298/210839653-5ea0e715-668f-4e50-9ce6-ade534e5e7b5.png)
```

so basically - there should maybe be a better error raised?

**Comment by dstansby** [comment]
I haven't tried making a astropy-only example, but given the traceback I'm 95% sure that this is a `wcsaxes` issue.
